src/Microsoft.AspNet.Server.Kestrel.Https/HttpsConnectionFilter.cs

```diff
@@ -35,10 +43,68 @@ public HttpsConnectionFilter(X509Certificate2 cert, IConnectionFilter previous)
             {
-                var sslStream = new SslStream(context.Connection);
-                await sslStream.AuthenticateAsServerAsync(_cert);
+                SslStream sslStream;
+                if (_clientCertMode == ClientCertificateMode.NoCertificate)
+                {
+                    sslStream = new SslStream(context.Connection);
+                    await sslStream.AuthenticateAsServerAsync(_serverCert);
+                }
+                else
+                {
+                    sslStream = new SslStream(context.Connection, leaveInnerStreamOpen: false,
+                        userCertificateValidationCallback: (sender, certificate, chain, sslPolicyErrors) =>
+                        {
+                            if (sslPolicyErrors.HasFlag(SslPolicyErrors.RemoteCertificateNotAvailable))
+                            {
+                                return _clientCertMode != ClientCertificateMode.RequireCertificate;
+                            }
+
+
+                            if (_clientValidationCallback == null)
+                            {
+                                if (sslPolicyErrors != SslPolicyErrors.None)
+                                {
+                                    return false;
+                                }
+                            }
+#if DOTNET5_4
+                            // conversion X509Certificate to X509Certificate2 not supported
+                            // https://github.com/dotnet/corefx/issues/4510
+                            X509Certificate2 certificate2 = null;
+                            return false;
+#else
+                            X509Certificate2 certificate2 = certificate as X509Certificate2 ??
+                                                            new X509Certificate2(certificate);
+
+#endif
+                            if (_clientValidationCallback != null)
+                            {
+                                if (!_clientValidationCallback(certificate2, chain, sslPolicyErrors))
+                                {
+                                    return false;
+                                }
+                            }
+
+                            _clientCert = certificate2;
+                            return true;
+                        });
+                    await sslStream.AuthenticateAsServerAsync(_serverCert, clientCertificateRequired: true,
+                        enabledSslProtocols: SslProtocols.Tls12 | SslProtocols.Tls11 | SslProtocols.Tls,
+                        checkCertificateRevocation: false);
+                }
                 context.Connection = sslStream;
             }
         }
+
+        public void PrepareRequest(IFeatureCollection features)
+        {
+            _previous.PrepareRequest(features);
+
+            if (_clientCert != null)
+            {
+                features.Set<ITlsConnectionFeature>(
+                    new TlsConnectionFeature { ClientCertificate = _clientCert });
```

Comment:
Can I use TlsConnectionFilter from Microsoft.AspNet.Http.Features.Internal or do should there be a class in the assembly?